ui/framework/src/test/imodel-components/VisibilityTreeEventHandler.test.ts

```diff
@@ -0,0 +1,119 @@
+
+  const modelSourceMock = moq.Mock.ofType<TreeModelSource>();
+  const modelMock = moq.Mock.ofType<TreeModel>();
+  const nodeLoaderMock = moq.Mock.ofType<AbstractTreeNodeLoaderWithProvider<IPresentationTreeDataProvider>>();
+  const dataProviderMock = moq.Mock.ofType<IPresentationTreeDataProvider>();
+  const selectionHandlerMock = moq.Mock.ofType<SelectionHandler>();
+
+  const changeVisibility = sinon.fake();
+  const getVisibilityStatus = sinon.stub();
+  const dispose = sinon.fake();
+  const onVisibilityChange = new BeEvent<VisibilityChangeListener>();
+
+  const visibilityHandler: IVisibilityHandler = {
+    changeVisibility,
+    getVisibilityStatus,
+    onVisibilityChange,
+    dispose,
+  };
+
+  const testVisibilityStatus: VisibilityStatus = {
+    state: "visible",
+    isDisabled: false,
+  };
+
+  beforeEach(async () => {
+    modelSourceMock.reset();
+    modelMock.reset();
+    nodeLoaderMock.reset();
+    dataProviderMock.reset();
+    selectionHandlerMock.reset();
+    changeVisibility.resetHistory();
+    getVisibilityStatus.reset();
+    dispose.resetHistory();
+
+    getVisibilityStatus.returns(testVisibilityStatus);
+    modelMock.setup((x) => x.getNode(moq.It.isAny())).returns(() => createSimpleTreeModelNode());
+    modelSourceMock.setup((x) => x.onModelChanged).returns(() => new BeUiEvent<[TreeModel, TreeModelChanges]>());
+    modelSourceMock.setup((x) => x.getModel()).returns(() => modelMock.object);
+    nodeLoaderMock.setup((x) => x.dataProvider).returns(() => dataProviderMock.object);
+    nodeLoaderMock.setup((x) => x.modelSource).returns(() => modelSourceMock.object);
+  });
+
+  const createHandler = (partialProps?: Partial<VisibilityTreeEventHandlerParams>): VisibilityTreeEventHandler => {
+    if (!partialProps)
+      partialProps = {};
+    const props: VisibilityTreeEventHandlerParams = {
+      visibilityHandler: partialProps.visibilityHandler || undefined,
+      nodeLoader: partialProps.nodeLoader || nodeLoaderMock.object,
+      selectionHandler: partialProps.selectionHandler || selectionHandlerMock.object,
+    };
+    return new VisibilityTreeEventHandler(props);
+  };
+
+  describe("onChangeVisibility", () => {
+    it("calls getVisibilityStatus() only on the nodes, whose visibility status is not known when onChangeVisibility event is raised with only visibilityStatus", async () => {
+      const visibilityStatus: Map<string, VisibilityStatus> = new Map([
+        ["testId2", testVisibilityStatus],
+      ]);
+
+      const treeModelNodes = [
+        createSimpleTreeModelNode("testId1"),
+        createSimpleTreeModelNode("testId2"),
+        createSimpleTreeModelNode("testId3"),
+      ];
+
+      modelMock.setup((x) => x.iterateTreeModelNodes()).returns(() => treeModelNodes[Symbol.iterator]());
+
+      await using(createHandler({ visibilityHandler }), async (_) => {
+        onVisibilityChange.raiseEvent(undefined, visibilityStatus);
+      });
+      await TestUtils.flushAsyncOperations();
```

Comment:
I would suggest moving this inside `using` after event is raised. 

Alse reset `getVisibilityStatus` history before raising `onVisibilityChange` event in order to avoid counting calls from constructor. You will probably need to add `flushAsyncOperations` before clearing history and raising event as constructor will have floating promise.

Reply:
Sure, updated